### PR TITLE
Favor #public_send over #send using Rubocop's cop

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

Devs keep using `#send` although that method does not preserve
private/protected visibility. Watching after this turned out to be quite
time-consuming while doing code review.

Currently, the Style/Send cop doesn't enforce `#public_send` however
(that's what we want). It simply discourages the use of #send. See
https://github.com/rubocop-hq/rubocop/pull/2081#issuecomment-292251650
for details. So a new entry on the Code Conventions doc has been added
to overcome this limitation:
https://github.com/openfoodfoundation/openfoodnetwork/wiki/Code-Conventions#prefer-public_send-over-send

#### What should we test?
A smoke test, including:

* Product import
* Subscriptions
* Uploading a new logo from Configuration > Contents
* Inventory items (from `admin/variant_overrides` URL)
* Capture (or any other modification to the payment) an order's payment (from `admin/orders/:order_id/payments` URL)
* Enterprise fees

Although very little, these features are affected by this change.

#### Release notes
Ensure devs use `#public_send` over `#send` by enabling a Rubocop cop.

Changelog Category: Added

### CODE REVIEW

**Comment by luisramos0** [comment]
There are 50 warnings, this means that when a dev touches one of these files in a PR, they will have to fix it to the new convention. Correct?

**Comment by sauloperez** [comment]
AFAIK that Cop doesn't have autocorrect but I can try to fix them

File: app/controllers/shop_controller.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def changeable_orders_alert
   private
 
   def filtered_json(products_json)
-    if applicator.send(:rules).any?
+    if applicator.rules.any?
```
Comment (sauloperez) [inline app/controllers/shop_controller.rb]: I made this one public. If we use it from outside, it's part of applicator's public API

File: app/controllers/spree/admin/payment_methods_controller_decorator.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@ module Admin
       # Only show payment methods that user has access to and sort by distributor name
       # ! Redundant code copied from Spree::Admin::ResourceController with modifications marked
       def collection
-        return parent.send(controller_name) if parent_data.present?
+        return parent.controller_name if parent_data.present?
```
Comment (sauloperez) [inline app/controllers/spree/admin/payment_methods_controller_decorator.rb]: [controller_name](https://apidock.com/rails/v3.2.13/ActionController/Metal/controller_name/class) is a public method

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ def ng_form_for(name, *args, &block)
   # spree.foo_path in any view rendered from non-spree-namespaced controllers.
   def method_missing(method, *args, &block)
     if (method.to_s.end_with?('_path') || method.to_s.end_with?('_url')) && spree.respond_to?(method)
-      spree.send(method, *args)
+      spree.public_send(method, *args)
```
Comment (sauloperez) [inline app/helpers/application_helper.rb]: Rails' URL helpers are public

File: app/models/model_set.rb [hunk 4]
```diff
@@ -12,7 +12,7 @@ def initialize(klass, collection, attributes={}, reject_if=nil, delete_if=nil)
     @collection = attributes[:collection] if attributes[:collection]
 
     attributes.each do |name, value|
-      send("#{name}=", value)
+      public_send("#{name}=", value)
```
Comment (sauloperez) [inline app/models/model_set.rb]: the setters of a model are meant to be public

File: app/models/proxy_order.rb [hunk 5]
```diff
@@ -41,7 +41,7 @@ def resume
     return false unless order_cycle.orders_close_at.andand > Time.zone.now
     transaction do
       update_column(:canceled_at, nil)
-      order.send('resume') if order
+      order.resume if order
```
Comment (sauloperez) [inline app/models/proxy_order.rb]: these two methods are public

File: app/services/order_syncer.rb [hunk 6]
```diff
@@ -83,7 +83,7 @@ def relevant_address_attrs
 
   def addresses_match?(order_address, subscription_address)
     relevant_address_attrs.all? do |attr|
-      order_address[attr] == subscription_address.send("#{attr}_was") ||
+      order_address[attr] == subscription_address.public_send("#{attr}_was") ||
```
Comment (sauloperez) [inline app/services/order_syncer.rb]: AR's dirty attributes API is public

File: lib/discourse/single_sign_on.rb [hunk 7]
```diff
@@ -42,7 +42,7 @@ def self.parse(payload, sso_secret = nil)
         if BOOLS.include? k
           val = ["true", "false"].include?(val) ? val == "true" : nil
         end
-        sso.send("#{k}=", val)
+        sso.public_send("#{k}=", val)
```
Comment (sauloperez) [inline lib/discourse/single_sign_on.rb]: each of these K are defined as public attr_accessor in line 11

File: lib/open_food_network/paperclippable.rb [hunk 8]
```diff
@@ -4,18 +4,18 @@
...
   module Paperclippable
     def self.included(base)
-      base.send :extend, ActiveModel::Naming
-      base.send :extend, ActiveModel::Callbacks
-      base.send :include, ActiveModel::Validations
-      base.send :include, Paperclip::Glue
+      base.extend(ActiveModel::Naming)
+      base.extend(ActiveModel::Callbacks)
+      base.include(ActiveModel::Validations)
+      base.include(Paperclip::Glue)
```
Comment (sauloperez) [inline lib/open_food_network/paperclippable.rb]: `extend` and `include` are public Module methods in Ruby 2.1.5 https://ruby-doc.org/core-2.1.5/Module.html#method-i-include

File: lib/open_food_network/paperclippable.rb [hunk 9]
```diff
@@ -4,18 +4,18 @@
...
 
       # Paperclip required callbacks
-      base.send :define_model_callbacks, :save, only: [:after]
-      base.send :define_model_callbacks, :commit, only: [:after]
-      base.send :define_model_callbacks, :destroy, only: [:before, :after]
+      base.define_model_callbacks(:save, only: [:after])
```
Comment (sauloperez) [inline lib/open_food_network/paperclippable.rb]: it is public https://apidock.com/rails/ActiveModel/Callbacks/define_model_callbacks

File: lib/spree/api/testing_support/setup.rb [hunk 10]
```diff
@@ -20,7 +20,9 @@ def sign_in_as_enterprise_user!(enterprises)
           let!(:current_api_user) do
             user = create(:user)
             user.spree_roles = []
-            enterprises.each { |e| user.enterprise_roles.create(enterprise: send(e)) }
+            enterprises.each do |enterprise|
+              user.enterprise_roles.create(enterprise: enterprise)
```
Comment (sauloperez) [inline lib/spree/api/testing_support/setup.rb]: no point for a send when `enterprise` is a var

**Comment by sauloperez** [comment]
Thanks for the heads-up. I think this now deserves a smoke test.

**Comment by Matt-Yorkley** [comment]
Looks like the tests are smoking already... :trollface:

File: app/controllers/spree/admin/payment_methods_controller_decorator.rb (hunk 2 above)
Comment (mkllnk) [reply app/controllers/spree/admin/payment_methods_controller_decorator.rb]: The original did not call `send(:controller_name)`. Doesn't it call a method with the same name as the controller?

Comment (sauloperez) [reply app/controllers/spree/admin/payment_methods_controller_decorator.rb]: :hankey: yes, I went too fast on this one. Thanks for the heads-up!

**Comment by RachL** [comment]
FINALLY!
Sorry it took me so long ...
Everything looks good:
 https://docs.google.com/document/d/1FLZVVyGaH6ubjeM-bpXHzeN4GxUTsTVC6fADeCO1xGk/edit

**Pull request merged**